File: MyUnityApp/Assets/Location/SimulatedLocationSource.cs
using System;
using System.Collections;
using UnityEngine;

namespace ARSpeedHUD.Location
{
    /// <summary>
    /// 네이티브 Android 앱의 SimulatedLocationSource.kt를 C#으로 이식한 버전이다.
    /// 실제 하드웨어, GPS, XREAL 글래스 없이도 러너의 물리적으로 그럴듯한 GPS 위치
    /// 정보 스트림을 생성한다. 준비 구간 동안 속도가 증가한 뒤 느린 사인파 변동과
    /// 잡음을 더해 조깅 페이스 주변을 오가며, 진행 방향도 무작위로 변하게 해
    /// 구불구불한 러닝 경로를 흉내 낸다. Unity 에디터 안에서 세션과 HUD 로직을
    /// 완전히 구축하고 조정할 수 있게 해 준다.
    /// </summary>
    public class SimulatedLocationSource : MonoBehaviour, ILocationSource
    {
        public LocationSourceType Type => LocationSourceType.Simulator;
        public event Action<RunSample> SampleReceived;

        [SerializeField] private double startLat = 37.5665;
        [SerializeField] private double startLon = 126.9780;
        [SerializeField] private float tickSeconds = 0.5f;
        [SerializeField] private double basePaceMps = 2.6; // 약 9.4km/h의 가벼운 조깅

        private Coroutine _routine;

        public void StartSampling()
        {
            StopSampling();
            _routine = StartCoroutine(Sample());
        }

        public void StopSampling()
        {
            if (_routine != null) StopCoroutine(_routine);
            _routine = null;
        }

        private IEnumerator Sample()
        {
            double lat = startLat;
            double lon = startLon;
            double bearingDeg = UnityEngine.Random.Range(0f, 360f);
            double elapsed = 0;
            const double earthRadius = 6371000.0;

            while (true)
            {
                elapsed += tickSeconds;

                double warmup = Math.Min(elapsed / 30.0, 1.0);
                double drift = Math.Sin(elapsed / 45.0) * 0.5;
                double noise = UnityEngine.Random.Range(-0.15f, 0.15f);
                double speed = Math.Min(Math.Max(basePaceMps * warmup + drift + noise, 0.0), 5.5);

                bearingDeg = (bearingDeg + UnityEngine.Random.Range(-6f, 6f) + 360.0) % 360.0;

                double distanceMeters = speed * tickSeconds;
                double bearingRad = bearingDeg * Math.PI / 180.0;
                lat += (distanceMeters * Math.Cos(bearingRad) / earthRadius) * (180.0 / Math.PI);
                lon += (distanceMeters * Math.Sin(bearingRad) / (earthRadius * Math.Cos(lat * Math.PI / 180.0))) * (180.0 / Math.PI);

                SampleReceived?.Invoke(new RunSample
                {
                    TimestampMillis = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                    Latitude = lat,
                    Longitude = lon,
                    SpeedMps = (float)speed,
                    AccuracyMeters = 3.5f
                });

                yield return new WaitForSeconds(tickSeconds);
            }
        }
    }
}

File: MyARSpeedHUD/app/src/main/java/com/myarspeedhud/app/location/SimulatedLocationSource.kt
package com.myarspeedhud.app.location

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

/**
 * 실제 하드웨어 없이도 러너(러닝 중인 사용자)의 물리적으로 적합한 GPS 데이터 좌표 스트림을 생성합니다.
 * 덕분에 Android 에뮬레이터에서 HUD를 완벽하게 테스트할 수 있습니다.
 * GPX 경로 재생이나 adb emu geo 명령을 호출할 필요도 없고, 실제 스마트 글래스 기기도 필요하지 않습니다.
 * 속도 보정, 페이스 계산, 누적 거리, HUD 렌더링에 이르는 전체 파이프라인을 스스로 처음부터 끝까지 구동합니다.
 * 속도는 짧은 웜업(준비 운동) 단계를 거치며 서서히 증가한 뒤, 약간의 노이즈와 느린 사인파(sinusoidal)를 이용하여 유동적으로 변합니다.
 * 방향(heading) 역시 무작위로 변경되면서 구불구불한 러닝 코스를 달리는 듯한 환경을 시뮬레이션 합니다.
 */
class SimulatedLocationSource(
    private val startLat: Double = 37.5665,
    private val startLon: Double = 126.9780,
    private val tickMillis: Long = 500L,
    private val basePaceMps: Double = 2.6 // ~9.4 km/h easy jog
) : LocationSource {
    override val type = LocationSourceType.SIMULATOR

    override fun samples(): Flow<RunSample> = flow {
        var lat = startLat
        var lon = startLon
        var bearingDeg = Random.nextDouble(0.0, 360.0)
        var elapsedMs = 0L
        val dtSeconds = tickMillis / 1000.0
        val earthRadius = 6371000.0

        while (true) {
            elapsedMs += tickMillis
            val tSec = elapsedMs / 1000.0

            val warmup = (tSec / 30.0).coerceAtMost(1.0)
            val drift = sin(tSec / 45.0) * 0.5
            val noise = Random.nextDouble(-0.15, 0.15)
            val speed = (basePaceMps * warmup + drift + noise).coerceIn(0.0, 5.5)

            bearingDeg = (bearingDeg + Random.nextDouble(-6.0, 6.0) + 360.0) % 360.0

            val distanceMeters = speed * dtSeconds
            val bearingRad = Math.toRadians(bearingDeg)
            lat += (distanceMeters * cos(bearingRad) / earthRadius) * (180.0 / PI)
            lon += (distanceMeters * sin(bearingRad) / (earthRadius * cos(Math.toRadians(lat)))) * (180.0 / PI)

            emit(
                RunSample(
                    timestampMillis = System.currentTimeMillis(),
                    latitude = lat,
                    longitude = lon,
                    speedMps = speed.toFloat(),
                    accuracyMeters = 3.5f
                )
            )
            delay(tickMillis)
        }
    }
}
